File: apps/macos/Mollotov/Handlers/NavigationHandler.swift
import Foundation

/// Handles navigate, back, forward, reload, getCurrentUrl.
struct NavigationHandler {
    let context: HandlerContext

    func register(on router: Router) {
        router.register("navigate") { body in await navigate(body) }
        router.register("back") { _ in await back() }
        router.register("forward") { _ in await forward() }
        router.register("reload") { _ in await reload() }
        router.register("get-current-url") { _ in await getCurrentUrl() }
    }

    @MainActor
    private func navigate(_ body: [String: Any]) async -> [String: Any] {
        guard let urlString = body["url"] as? String,
              let url = URL(string: urlString),
              context.renderer != nil else {
            return errorResponse(code: "INVALID_URL", message: "Missing or invalid URL")
        }
        let start = CFAbsoluteTimeGetCurrent()
        context.load(url: url)

        for _ in 0..<100 {
            try? await Task.sleep(nanoseconds: 100_000_000)
            if !context.isLoadingPage { break }
        }

        let loadTime = Int((CFAbsoluteTimeGetCurrent() - start) * 1000)
        return successResponse([
            "url": context.currentURL?.absoluteString ?? urlString,
            "title": context.currentTitle,
            "loadTime": loadTime,
        ])
    }

    @MainActor
    private func back() async -> [String: Any] {
        guard context.renderer != nil else { return errorResponse(code: "NO_WEBVIEW", message: "No WebView") }
        context.goBack()
        try? await Task.sleep(nanoseconds: 500_000_000)
        return successResponse(["url": context.currentURL?.absoluteString ?? "", "title": context.currentTitle])
    }

    @MainActor
    private func forward() async -> [String: Any] {
        guard context.renderer != nil else { return errorResponse(code: "NO_WEBVIEW", message: "No WebView") }
        context.goForward()
        try? await Task.sleep(nanoseconds: 500_000_000)
        return successResponse(["url": context.currentURL?.absoluteString ?? "", "title": context.currentTitle])
    }

    @MainActor
    private func reload() async -> [String: Any] {
        guard context.renderer != nil else { return errorResponse(code: "NO_WEBVIEW", message: "No WebView") }
        let start = CFAbsoluteTimeGetCurrent()
        context.reloadPage()
        for _ in 0..<100 {
            try? await Task.sleep(nanoseconds: 100_000_000)
            if !context.isLoadingPage { break }
        }
        let loadTime = Int((CFAbsoluteTimeGetCurrent() - start) * 1000)
        return successResponse(["url": context.currentURL?.absoluteString ?? "", "title": context.currentTitle, "loadTime": loadTime])
    }

    @MainActor
    private func getCurrentUrl() async -> [String: Any] {
        guard context.renderer != nil else { return errorResponse(code: "NO_WEBVIEW", message: "No WebView") }
        return ["url": context.currentURL?.absoluteString ?? "", "title": context.currentTitle]
    }
}

File: native/engine-chromium-desktop/src/handlers/navigation_handler.cpp
#include "navigation_handler.h"

namespace mollotov {

NavigationHandler::NavigationHandler(DesktopHandlerRuntime runtime)
    : runtime_(std::move(runtime)) {}

void NavigationHandler::Register(DesktopRouter& router) const {
  router.Register("navigate", [this](const nlohmann::json& params) { return Navigate(params); });
  router.Register("back", [this](const nlohmann::json& params) { return Back(params); });
  router.Register("forward", [this](const nlohmann::json& params) { return Forward(params); });
  router.Register("reload", [this](const nlohmann::json& params) { return Reload(params); });
  router.Register("get-current-url",
                  [this](const nlohmann::json&) { return GetCurrentUrl(); });
}

nlohmann::json NavigationHandler::Navigate(const nlohmann::json& params) const {
  try {
    const std::string url = RequireString(params, "url");
    const std::int64_t started = NowMillis();
    HandlerContext& context = RequireHandlerContext(runtime_);
    context.Renderer()->LoadUrl(url);

    const std::string current_url = context.Renderer()->CurrentUrl();
    const std::string title = context.Renderer()->CurrentTitle();
    if (runtime_.history_store != nullptr) {
      runtime_.history_store->Record(current_url.empty() ? url : current_url, title);
    }
    return SuccessResponse({
        {"url", current_url.empty() ? url : current_url},
        {"title", title},
        {"loadTime", static_cast<int>(NowMillis() - started)},
    });
  } catch (const std::invalid_argument& exception) {
    return InvalidParams(exception.what());
  }
}

nlohmann::json NavigationHandler::Back(const nlohmann::json&) const {
  HandlerContext& context = RequireHandlerContext(runtime_);
  context.Renderer()->GoBack();
  return SuccessResponse({
      {"url", context.Renderer()->CurrentUrl()},
      {"title", context.Renderer()->CurrentTitle()},
  });
}

nlohmann::json NavigationHandler::Forward(const nlohmann::json&) const {
  HandlerContext& context = RequireHandlerContext(runtime_);
  context.Renderer()->GoForward();
  return SuccessResponse({
      {"url", context.Renderer()->CurrentUrl()},
      {"title", context.Renderer()->CurrentTitle()},
  });
}

nlohmann::json NavigationHandler::Reload(const nlohmann::json&) const {
  const std::int64_t started = NowMillis();
  HandlerContext& context = RequireHandlerContext(runtime_);
  context.Renderer()->Reload();
  return SuccessResponse({
      {"url", context.Renderer()->CurrentUrl()},
      {"loadTime", static_cast<int>(NowMillis() - started)},
  });
}

nlohmann::json NavigationHandler::GetCurrentUrl() const {
  HandlerContext& context = RequireHandlerContext(runtime_);
  return {
      {"url", context.Renderer()->CurrentUrl()},
      {"title", context.Renderer()->CurrentTitle()},
  };
}

}  // namespace mollotov
